File: ScaleformUI_Csharp/Scaleforms/RankBar/RankBarHandler.cs
using CitizenFX.Core;
using static CitizenFX.FiveM.Native.Natives;

namespace ScaleformUI.Scaleforms
{
    public class RankBarHandler
    {
        private const int HUD_COMPONENT_ID = 19;

        private HudColor _rankBarColor = HudColor.HUD_COLOUR_FREEMODE;

        /// <summary>
        /// Set the color of the Rank Bar when displayed
        /// </summary>
        public HudColor Color
        {
            get => _rankBarColor;
            set
            {
                _rankBarColor = value;
            }
        }

        public RankBarHandler() { }

        public async Task Load()
        {
            int timeout = 1000;

            int start = Main.GameTime;
            RequestHudScaleform(HUD_COMPONENT_ID);
            while (!HasHudScaleformLoaded(HUD_COMPONENT_ID) && Main.GameTime - start < timeout)
            {
                await BaseScript.Delay(0);
            }
        }

        /// <summary>
        /// Shows the rank bar
        /// </summary>
        /// <param name="limitStart">Floor value of experience e.g. 0</param>
        /// <param name="limitEnd">Ceiling value of experience e.g. 800</param>
        /// <param name="previousValue">Previous Experience value </param>
        /// <param name="currentValue">New Experience value</param>
        /// <param name="currentRank">Current rank</param>
        public async void SetScores(int limitStart, int limitEnd, int previousValue, int currentValue, int currentRank)
        {
            await Load();

            // Color has to be set else it will be white by default
            BeginScaleformMovieMethodHudComponent(HUD_COMPONENT_ID, "SET_COLOUR");
            PushScaleformMovieFunctionParameterInt((int)_rankBarColor);
            EndScaleformMovieMethod();

            // this will set an update the score
            BeginScaleformMovieMethodHudComponent(HUD_COMPONENT_ID, "SET_RANK_SCORES");
            PushScaleformMovieFunctionParameterInt(limitStart);
            PushScaleformMovieFunctionParameterInt(limitEnd);
            PushScaleformMovieFunctionParameterInt(previousValue);
            PushScaleformMovieFunctionParameterInt(currentValue);
            PushScaleformMovieFunctionParameterInt(currentRank);
            PushScaleformMovieFunctionParameterInt(100);
            EndScaleformMovieMethod();
        }

        public void Remove()
        {
            if (HasHudScaleformLoaded(HUD_COMPONENT_ID))
            {
                BeginScaleformScriptHudMovieMethod(HUD_COMPONENT_ID, "REMOVE");
                EndScaleformMovieMethod();
            }
        }

        public async void OverrideAnimationSpeed(int speed = 1000)
        {
            await Load();
            BeginScaleformScriptHudMovieMethod(HUD_COMPONENT_ID, "OVERRIDE_ANIMATION_SPEED");
            PushScaleformMovieFunctionParameterInt(speed);
            EndScaleformMovieMethod();
        }

        public async void OverrideOnscreenDuration(int duration = 4000)
        {
            await Load();
            BeginScaleformScriptHudMovieMethod(HUD_COMPONENT_ID, "OVERRIDE_ONSCREEN_DURATION");
            PushScaleformMovieFunctionParameterInt(duration);
            EndScaleformMovieMethod();
        }
    }
}

File: ScaleformUI_Lua/src/scaleforms/Rankbar/RankbarHandler.lua
RankbarHandler = setmetatable({}, RankbarHandler)
RankbarHandler.__index = RankbarHandler
RankbarHandler.__call = function()
    return "RankbarHandler"
end

---@class RankbarHandler
---@field public Load fun():promise
---@field public SetScores fun(limitStart:number, limitEnd:number, previousValue:number, currentValue:number, currentRank:number):nil
---@field public SetColour fun(rankBarColor:number):nil
---@field public Remove fun():nil
---@field public OverrideAnimationSpeed fun(speed:number):nil
---@field public OverrideOnscreenDuration fun(duration:number):nil

local HUD_COMPONENT_ID = 19
local _rankBarColor = 116

---Loads the rankbar scaleform movie
---@return promise
function RankbarHandler:Load()
    local p = promise.new()

    if HasScaleformScriptHudMovieLoaded(HUD_COMPONENT_ID) then
        p:resolve()
        return p
    end

    RequestScaleformScriptHudMovie(HUD_COMPONENT_ID)
    local timeout = 1000
    local start = GlobalGameTimer
    while not HasScaleformScriptHudMovieLoaded(HUD_COMPONENT_ID) and GlobalGameTimer - start < timeout do Citizen.Wait(0) end

    if HasScaleformScriptHudMovieLoaded(HUD_COMPONENT_ID) then
        p:resolve()
    else
        p:reject()
    end

    return p
end

---Set the scores for the rankbar
---@param limitStart number
---@param limitEnd number
---@param previousValue number
---@param currentValue number
---@param currentRank number
function RankbarHandler:SetScores(limitStart, limitEnd, previousValue, currentValue, currentRank)
    self:Load():next(function()
            BeginScaleformScriptHudMovieMethod(HUD_COMPONENT_ID, "SET_COLOUR")
            ScaleformMovieMethodAddParamInt(_rankBarColor)
            EndScaleformMovieMethod()

            BeginScaleformScriptHudMovieMethod(HUD_COMPONENT_ID, "SET_RANK_SCORES")
            ScaleformMovieMethodAddParamInt(limitStart)
            ScaleformMovieMethodAddParamInt(limitEnd)
            ScaleformMovieMethodAddParamInt(previousValue)
            ScaleformMovieMethodAddParamInt(currentValue)
            ScaleformMovieMethodAddParamInt(currentRank)
            EndScaleformMovieMethod()
        end,
        function()
            print("RankbarHandler:Load() failed")
        end)
end

---Set the colour of the rankbar
---@param rankBarColor HudColours
function RankbarHandler:SetColour(rankBarColor)
    _rankBarColor = rankBarColor
end

---Remove the rankbar from the screen
function RankbarHandler:Remove()
    if HasScaleformScriptHudMovieLoaded(HUD_COMPONENT_ID) then
        BeginScaleformScriptHudMovieMethod(HUD_COMPONENT_ID, "REMOVE");
        EndScaleformMovieMethod();
    end
end

---Override the animation speed of the rankbar
---@param speed number -- milliseconds (default 1000)
function RankbarHandler:OverrideAnimationSpeed(speed)
    self:Load():next(function()
            BeginScaleformScriptHudMovieMethod(HUD_COMPONENT_ID, "OVERRIDE_ANIMATION_SPEED");
            ScaleformMovieMethodAddParamInt(speed);
            EndScaleformMovieMethod();
        end,
        function()
            print("RankbarHandler:Load() failed")
        end)
end

---Override the onscreen duration of the rankbar
---@param duration number -- milliseconds (default 4000)
function RankbarHandler:OverrideOnscreenDuration(duration)
    self:Load():next(function()
            BeginScaleformScriptHudMovieMethod(HUD_COMPONENT_ID, "OVERRIDE_ONSCREEN_DURATION");
            ScaleformMovieMethodAddParamInt(duration);
            EndScaleformMovieMethod();
        end,
        function()
            print("RankbarHandler:Load() failed")
        end)
end
